test/performance/odp_timer_stress.c

```diff
@@ -0,0 +1,1156 @@
+	odp_fract_u64_t hz;
+	double hz_d, min_hz_d, max_hz_d;
+	odp_pool_capability_t pool_capa;
+	odp_shm_capability_t shm_capa;
+	uint64_t req_shm_sz;
+
+	if (opts->mode != SINGLE_SHOT && opts->mode != PERIODIC && opts->mode != CANCEL) {
+		ODPH_ERR("Invalid timer mode: %u\n", opts->mode);
+		return PRS_NOK;
+	}
+
+	if (opts->policy != SHARED_TMR && opts->policy != PRIV_TMR) {
+		ODPH_ERR("Invalid pool policy: %d\n", opts->policy);
+		return PRS_NOK;
+	}
+
+	if (opts->mode == CANCEL && opts->policy != PRIV_TMR) {
+		ODPH_ERR("Single shot with cancel mode supported only with worker-private "
+			 "timers\n");
+		return PRS_NOK;
+	}
+
+	ret = odp_timer_capability(opts->clk_src, &tmr_capa);
+
+	if (ret < -1) {
+		ODPH_ERR("Error querying timer capabilities\n");
+		return PRS_NOK;
+	}
+
+	if (ret == -1) {
+		ODPH_ERR("Invalid clock source: %d\n", opts->clk_src);
+		return PRS_NOK;
+	}
+
+	if (!tmr_capa.queue_type_sched) {
+		ODPH_ERR("Invalid queue support, scheduled completion queues not supported\n");
+		return PRS_NOK;
+	}
+
+	if (opts->res_ns < tmr_capa.highest_res_ns) {
+		ODPH_ERR("Invalid resolution: %" PRIu64 " ns (max: %" PRIu64 " ns)\n",
+			 opts->res_ns, tmr_capa.highest_res_ns);
+		return PRS_NOK;
+	}
+
+	if (opts->num_tmr == 0U) {
+		ODPH_ERR("Invalid number of timers: %u\n", opts->num_tmr);
+		return PRS_NOK;
+	}
+
+	max_workers = ODPH_MIN(MAX_WORKERS, (uint32_t)odp_cpumask_default_worker(NULL, 0));
+
+	if (opts->num_workers == 0U || opts->num_workers > max_workers) {
+		ODPH_ERR("Invalid worker count: %u (min: 1, max: %u)\n", opts->num_workers,
+			 max_workers);
+		return PRS_NOK;
+	}
+
+	(void)odp_cpumask_default_worker(&config->worker_mask, opts->num_workers);
+
+	req_tmr = opts->num_tmr * (opts->policy == PRIV_TMR ? opts->num_workers : 1U);
+
+	if (opts->mode == SINGLE_SHOT || opts->mode == CANCEL) {
+		if (tmr_capa.max_pools == 0U) {
+			ODPH_ERR("Single shot timers not supported\n");
+			return PRS_NOK;
+		}
+
+		if (tmr_capa.max_timers > 0U && req_tmr > tmr_capa.max_timers) {
+			ODPH_ERR("Invalid number of timers: %u (max: %u)\n", req_tmr,
+				 tmr_capa.max_timers);
+			return PRS_NOK;
+		}
+
+		config->res_capa.res_ns = opts->res_ns;
+
+		if (odp_timer_res_capability(opts->clk_src, &config->res_capa) < 0) {
+			ODPH_ERR("Error querying timer resolution capabilities\n");
+			return PRS_NOK;
+		}
+
+		if (opts->mode == CANCEL) {
+			if (odp_shm_capability(&shm_capa) < 0) {
+				ODPH_ERR("Error querying SHM capabilities");
+				return PRS_NOK;
+			}
+
+			/* One block for program configuration, one block divided between
+			 * workers. */
+			req_shm = 2U;
+
+			if (req_shm > shm_capa.max_blocks) {
+				ODPH_ERR("Invalid SHM block count support: %u (max: %u)\n",
+					 req_shm, shm_capa.max_blocks);
+				return PRS_NOK;
+			}
+
+			/* Dimensioned so that each structure will always start at a cache line
+			 * boundary. */
+			req_shm_sz = ODP_CACHE_LINE_SIZE *
+				     ROUNDUP_DIV(sizeof(tmr_hdls_t), ODP_CACHE_LINE_SIZE) *
```

Comment:
Could be done also with `ODP_CACHE_LINE_ROUNDUP()` macro.

Reply:
Thanks, forgot that we have that available publicly, v6.